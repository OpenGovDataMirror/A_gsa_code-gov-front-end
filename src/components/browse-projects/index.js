import BrowseProjectsContainer from './browse-projects.container'

export default BrowseProjectsContainer
