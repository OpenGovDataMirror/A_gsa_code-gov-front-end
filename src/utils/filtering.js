export function isChecked(item) {
  return typeof item === 'object' && item.checked
}