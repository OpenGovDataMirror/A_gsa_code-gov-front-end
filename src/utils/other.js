/* global PUBLIC_PATH */
/* global SITE_CONFIG */
import { find, lower, startsWith, trim } from '@code.gov/cautious'
import get from 'lodash.get'

export const falses = [undefined, null, 'null', 'None', 'Null', 'NULL', '', 'False', 'false']

export function isFalse(input) {
  return falses.includes(input)
}

export function pathMatch(url, match) {
  return (
    startsWith(url, `./${match}`) || startsWith(url, `/${match}/`) || startsWith(url, `${match}/`)
  )
}

export function updatedRecursivePath(thing) {
  if (typeof thing === 'object') {
    for (const key in thing) {
      const subvalue = thing[key]
      if (typeof subvalue === 'string') {
        // eslint-disable-next-line no-use-before-define
        thing[key] = adjustAssetPath(subvalue)
      }
    }
  }
  return thing
}

export function adjustAssetPath(thing) {
  if (pathMatch(thing, 'assets')) {
    const pattern = /.?\/?assets\//
    const newAssetPath = `${PUBLIC_PATH}assets/`
    return thing.replace(pattern, newAssetPath)
  }

  return updatedRecursivePath(thing)
}

export function getConfigValue(path) {
  if (SITE_CONFIG) {
    let value = get(SITE_CONFIG, path)
    if (typeof value === 'string') {
      try {
        value = adjustAssetPath(value)
      } catch (error) {
        console.error("couldn't check if the following started with assets", [value])
      }
    } else if (Array.isArray(value)) {
      value = value.map(item => {
        if (typeof item === 'object') {
          for (const subkey in item) {
            const subvalue = item[subkey]
            if (typeof subvalue === 'string') {
              item[subkey] = adjustAssetPath(subvalue)
            }
          }
        }
        return item
      })
    }
    if (!value) {
      console.warn(`We weren't able to find the value for ${path} in your code-gov-config.json file.
      You can consult examples under the folder found at config/site/examples
      or consult our developer documentation here: https://open.gsa.gov/api/codedotgov/`)
    }
    return value
  }
  return null
}

export function normalize(input) {
  return trim(lower(input))
}

export function setScrollDepth(scrollDepth) {
  if (document && document.documentElement && document.documentElement.scrollTop) {
    document.documentElement.scrollTop = scrollDepth
  }
  if (document.body && document.body.scrollTop) {
    document.body.scrollTop = scrollDepth
  }
}

/* runs when each page component is loaded */
export function refreshView() {
  window.scrollTo(0, 0)
  setScrollDepth(0)
  if (document.activeElement) {
    document.activeElement.blur()
  }
}

export function scrollToTopOfResults() {
  console.log('starting scrollToTopOfResults')
  const headerHeight = document.querySelector('.menu-banner-header').clientHeight
  const bannerHeight = document.querySelector('.usa-section').clientHeight
  const scrollDepth = headerHeight + bannerHeight
  const scrollOptionsSupported = document.documentElement.style.scrollBehavior !== undefined
  if (scrollOptionsSupported) {
    window.scrollTo({
      behavior: 'smooth',
      left: 0,
      top: scrollDepth
    })
  } else {
    window.scrollTo(0, scrollDepth)
  }
  setScrollDepth(scrollDepth)
}

/* gets a set of values given a path */
export function getSet(items, path) {
  const results = new Set()
  if (Array.isArray(items) && typeof path === 'string') {
    items.forEach(item => {
      const value = get(item, path)
      if (value && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(subitem => {
            if (isFalse(subitem) === false) {
              results.add(subitem)
            }
          })
        } else if (isFalse(value) === false) {
          results.add(value)
        }
      }
    })
  }
  return results
}

export function getLowerSet(items, path) {
  const results = new Set()
  if (Array.isArray(items) && typeof path === 'string') {
    items.forEach(item => {
      const value = get(item, path)
      if (value && value !== '') {
        if (Array.isArray(value)) {
          value.forEach(subitem => {
            if (isFalse(subitem) === false) {
              results.add(subitem.toLowerCase())
            }
          })
        } else if (isFalse(value) === false) {
          results.add(value.toLowerCase())
        }
      }
    })
  }
  return results
}

export function getFilterData(key, path, currentSearchResults, filters) {
  if (currentSearchResults && filters && filters[key]) {
    const names = getLowerSet(currentSearchResults.repos, path)
    return filters[key].filter(
      ({ name, value }) => names.has(normalize(name)) || names.has(normalize(value))
    )
  }
}

export function hasLicense(repo) {
  return (
    repo.permissions &&
    Array.isArray(repo.permissions.licenses) &&
    repo.permissions.licenses.length > 0
  )
}

export function now() {
  return new Date().getTime()
}

export function getFilterValuesFromParamsByCategory(params = {}, category) {
  console.log('other-getFilter: category', category)
  return params.filters
    .filter(entry => entry.category === category)
    .map(entry => entry.value)
    .map(value => (typeof value === 'string' ? value.toLowerCase().trim() : value))
}

export function getFilterTags(params, filters) {
  return params.filters
    .map(({ category, modified, value }) => {
      const normalizedValue = value.toLowerCase()
      const found = find(
        get(filters, category),
        item => item.value.toLowerCase() === normalizedValue
      )
      let title = 'loading'
      if (found) {
        /* eslint-disable */
        if (found.name) title = found.name
        if (found.value) value = found.value
        /* eslint-enable */
      }
      return { category, modified, value, title }
    })
    .sort((a, b) => Math.sign(a.modified - b.modified))
}

export function loadScript(src) {
  console.log('starting loadScript with', src)
  return new Promise((resolve, _reject) => {
    const script = document.createElement('script')
    script.src = src
    script.onload = resolve
    document.body.appendChild(script)
    console.log('appended:', script, 'to body')
  })
}

export function fillFilters(keys, params, result) {
  keys.forEach(key => {
    if (Array.isArray(params[key])) {
      params[key].forEach(value => {
        result.filters.push({
          category: key,
          value,
          modified: now()
        })
      })
    }
  })
}

export function onHomePage() {
  return window.location.pathname === PUBLIC_PATH
}

/* I'd prefer to use fetch but IE polyfilling is complicated */
export function getText(url) {
  return new Promise(resolve => {
    const xhr = new XMLHttpRequest()
    xhr.onreadystatechange = function() {
      if (xhr.readyState === 4) {
        resolve(xhr.response)
      }
    }
    xhr.open('GET', url, true)
    xhr.send('')
  })
}

export function getJSON(url) {
  return getText(url).then(JSON.parse)
}

// https://stackoverflow.com/questions/1500260/detect-urls-in-text-with-javascript
export function urlify(text) {
  const urlRegex = /(https?:\/\/[^\s']+)/g
  return text.replace(urlRegex, url => `<a href="${url}" target="_blank">${url}</a>`)
}

export function prettify(text) {
  // convert (1) cost: to \n\t(1) <b>cost</b>:
  const bulletRegex = /\(\d{1,2}\) [A-Za-z]{1,25}:/g
  return urlify(text).replace(
    bulletRegex,
    match => `<br/>${match.replace(/[A-Za-z]{1,25}/, name => `<b>${name}</b>`)}`
  )
}

export const isHomepage = window.location.pathname === PUBLIC_PATH
