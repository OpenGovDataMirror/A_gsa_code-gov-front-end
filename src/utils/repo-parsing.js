import get from 'lodash.get'
import { filter } from '@code.gov/cautious'

export function getDisplayTextForUsageType(repo) {
  let text = ''
  const usageType = get(repo, 'permissions.usageType')
  if (usageType === 'openSource') {
    text = 'Open Source'
  } else if (usageType === 'governmentWideReuse') {
    text = 'Gov-wide Reuse'
  } else if (usageType) {
    text = usageType
  } else {
    text = 'unknown'
  }
  return text
}

export function getLicenseName(repo) {
  return get(repo, 'permissions.licenses[0].name')
}

export function getLaborHours(repo) {
  try {
    const laborHours = Number(get(repo, 'contact.laborHours'))
    if (laborHours > 0) {
      return laborHours
    }
  } catch (error) {
    console.warn(error)
  }
}

export function parseLanguages(repo) {
  const languages = get(repo, 'languages')
  if (Array.isArray(languages)) {
    return filter(languages, Boolean)
  }
  return []
}

export function parseEmail(repo) {
  const email = get(repo, 'contact.email')
  if (email) {
    return email
  }
}

export function parseRepositoryURL(repo) {
  let url = get(repo, 'repositoryURL')

  // some repos return 'null'
  if (!url || url === 'null') {
    return ''
  }

  if (url.startsWith('git://github.com/')) {
    url = url.replace('git://github.com/', 'https://github.com/')
  } else if (url.startsWith('git@github.com:')) {
    url = url.replace('git@github.com:', 'https://github.com/').replace('.git', '')
  } else if (url.startsWith('https://github.com') && url.endsWith('.git')) {
    url = url.replace('.git', '')
  }

  return url
}

export function parseTags(repo) {
  try {
    return get(repo, 'tags').filter(Boolean)
  } catch (error) {
    return []
  }
}

export function getDate(item) {
  return get(item, 'ghUpdatedAt') || get(item, 'date.lastModified')
}

export function getLastModifiedDateString(repo) {
  try {
    const dateLastModified = getDate(repo)
    if (dateLastModified) {
      return new Date(dateLastModified).toLocaleDateString('en-US')
    }
  } catch (error) {
    console.warn(error)
  }
}
