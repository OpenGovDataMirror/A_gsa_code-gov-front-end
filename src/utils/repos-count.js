export function getReposCount(total, messages, cssClasses = '') {
  let textContent = messages.default

  const len = Object.keys(messages).length - 1
  if (total >= len) {
    textContent = Object.values(messages)[len - 1]
  } else {
    Object.values(messages).forEach((item, key) => {
      if (key === total) {
        textContent = item
      }
    })
  }

  textContent = textContent.replace('[VALUE]', total)
  return `<p class="${cssClasses}">${textContent}</p>`
}
