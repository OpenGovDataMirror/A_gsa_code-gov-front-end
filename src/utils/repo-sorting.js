import { getDate } from './repo-parsing'

export function clean(string) {
  try {
    return string
      .toLowerCase()
      .replace(/-/g, ' ')
      .trim()
  } catch (error) {
    return string
  }
}

export function getTime(date) {
  try {
    if (date) {
      return new Date(date).getTime()
    }
    return -10e10
  } catch (error) {
    return -10e10
  }
}

export function sortByBestMatch(a, b) {
  return Number(a.searchScore) > Number(b.searchScore) ? -1 : 1
}

export function sortByDataQuality(a, b) {
  return Number(a.score) > Number(b.score) ? -1 : 1
}

export function sortByDate(a, b) {
  try {
    const aDate = getDate(a)
    const bDate = getDate(b)
    const aTime = getTime(aDate)
    const bTime = getTime(bDate)
    return Math.sign(bTime - aTime) || 0
  } catch (error) {
    console.error(error)
    return 0
  }
}

export function sortByName(a, b) {
  return clean(a.name) > clean(b.name) ? 1 : -1
}
