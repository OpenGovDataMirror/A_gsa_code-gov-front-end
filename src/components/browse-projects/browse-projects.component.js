import React from 'react'
import ReactHtmlParser from 'react-html-parser'
import Breadcrumbs from 'components/breadcrumbs'
import FilterBoxes from 'components/filter-boxes'
import FilterTags from 'components/filter-tags'
import Pagination from 'components/pagination'
import QualityPopover from 'components/quality-popover'
import QuickSearchBox from 'components/quick-search-box'
import SiteBanner from 'components/site-banner'
import SortSection from 'components/sort-section'
import RepoCard from 'components/repo-card'
import { refreshView, scrollToTopOfResults } from 'utils/other'
import { some } from '@code.gov/cautious'
import { getReposCount } from '../../utils/repos-count'

export default class BrowseProjects extends React.Component {
  componentDidMount() {
    refreshView()
    if (!this.props.filterData) this.props.saveFilterData()
  }

  onFilterBoxChange(category, values) {
    scrollToTopOfResults()
    this.props.onFilterBoxChange(category, values)
  }

  get repoCounter() {
    const { total } = this.props

    const messages = {
      default: 'Loading Repositories',
      '0': 'No Repositories',
      '1': '1 Repository',
      '2': `[VALUE] Repositories`
    }

    return getReposCount(
      total,
      messages,
      'repos-count tablet-lg:grid-col-9 grid-col-12 tablet:grid-col-4 font-heading-lg text-bold tablet:margin-y-105 margin-top-3 margin-bottom-0'
    )
  }

  get reposContainer() {
    if (some(this.props.repos)) {
      try {
        return (
          <div>
            <FilterTags filters={this.props.filterTags} onClick={::this.props.onFilterTagClick} />
            <div>
              <QualityPopover />
            </div>
            <ul className="usa-card-group padding-top-3">
              {this.props.repos.map(repo => (
                <RepoCard key={repo.repoID} repo={repo} />
              ))}
            </ul>
          </div>
        )
      } catch (error) {
        console.error('reposContainer error with this.props.repos', this.props.repos)
        throw error
      }
    }
  }

  updatePage(newPage) {
    scrollToTopOfResults()
    this.props.updatePage(newPage)
    document.getElementsByClassName('project-link')[0].focus()
  }

  render() {
    const numPages = Math.ceil(this.props.total / this.props.selectedPageSize)
    return (
      <main className="search-results-content" id="main-content">
        <SiteBanner title="Projects" />
        <Breadcrumbs crumbs={[{ text: 'Home', to: '/' }, { text: 'Projects' }]} />
        <div className="grid-container">
          <div className="grid-row grid-gap">
            <div className="margin-top-1 grid-col-12 tablet-lg:grid-col-3 tablet:grid-col-8">
              <QuickSearchBox />
            </div>
            {ReactHtmlParser(this.repoCounter)}
          </div>
        </div>
        <div className="grid-container">
          <div className="grid-row grid-gap">
            <div
              id="filter-boxes-section"
              className="tablet-lg:grid-col-3 tablet-lg:margin-top-4 margin-top-3"
            >
              <h2 className="tablet-lg:margin-bottom-4 margin-bottom-105">Filter</h2>

              <FilterBoxes
                boxes={this.props.boxes}
                config={[
                  ['Language', 'languages'],
                  ['Federal Agency', 'agencies'],
                  ['Licenses', 'licenses'],
                  ['Usage Types', 'usageTypes']
                ]}
                onFilterBoxChange={::this.onFilterBoxChange}
              />
            </div>
            <div id="filter-results-section" className=" tablet-lg:grid-col-9">
              <SortSection
                options={this.props.sortOptions}
                onSortChange={this.props.onSortChange}
              />
              <div className="card-list">
                {this.reposContainer}
                {numPages > 0 && (
                  <Pagination
                    count={this.props.total}
                    pagesize={this.props.selectedPageSize}
                    page={this.props.selectedPage}
                    updatePage={::this.updatePage}
                  />
                )}
              </div>
            </div>
          </div>
        </div>
      </main>
    )
  }
}
