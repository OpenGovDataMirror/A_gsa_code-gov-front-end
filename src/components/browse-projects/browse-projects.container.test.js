import saveFilterOptions from 'actions/save-filter-options'
import updateBrowseFilters from 'actions/update-browse-filters'
import updateBrowseParams from 'actions/update-browse-params'
import * as otherUtils from 'utils/other'
import { mapStateToProps, mapDispatchToProps } from 'components/browse-projects/browse-projects.container'

jest.mock('actions/save-filter-options')
jest.mock('actions/update-browse-filters')
jest.mock('actions/update-browse-params')

const props = {
  browseParams: {
    page: 'test-page',
    sort: 'test-sort',
    size: 'test-size',
    filters: [
      { category: 'agencies', value: 'v0', modified: otherUtils.now() },
      { category: 'agencies', value: 'v2', modified: otherUtils.now() },
      { category: 'languages', value: 'v3', modified: otherUtils.now() },
    ],
  },
  browseResults: {
    total: 2,
    repos: [{ repoId: 'repo-1' }, { repoId: 'repo-2' }],
  },
  filters: {
    agencies: [
      { name: 'a1', value: 'v1' },
      { name: 'a2', value: 'v2' },
    ],
    languages: [
      { name: 'l1', value: 'v3' },
    ],
    licenses: [],
    usageTypes: [],
  }
}

const dispatch = jest.fn()

describe('containers - BrowseProjects', () => {
  describe('mapStateToProps', () => {
    it('should map `boxes` based off `filters`, checked if a matching browseParam filter value', () => {
      const actual = mapStateToProps(props).boxes
      const expected = {
        agencies: [
          { name: 'a1', value: 'v1', checked: false }, // no matching browseParams value fpor 'v1'
          { name: 'a2', value: 'v2', checked: true },
        ],
        languages: [
          { name: 'l1', value: 'v3', checked: true },
        ],
        licenses: [],
        usageTypes: [],
      }
      expect(actual).toEqual(expected)
    })

    it('should map no `boxes` if no filters provided', () => {
      expect(mapStateToProps({ ...props, filters: undefined }).boxes).toEqual({})
    })

    it('should map `filterTags` based off `browseParams` and `filters`', () => {
      jest.spyOn(otherUtils, 'getFilterTags')
      mapStateToProps(props)
      expect(otherUtils.getFilterTags).toBeCalledWith(props.browseParams, props.filters)
    })

    it.each`
      selected
      ${'data_quality'}
      ${'a-z'}
      ${'last_updated'}
    `('should set the `sortOptions` value as $selected when selected', ({ selected }) => {
      const { sortOptions } = mapStateToProps({
        ...props,
        browseParams: { ...props.browseParams, sort: selected },
      })
      expect(sortOptions.find(x => x.selected).value).toBe(selected)
    })

    it('should return the correct properties', () => {
      expect(mapStateToProps(props)).toMatchSnapshot()
    })

    it('should default to 0 `total` if none provided in `browseResults`', () => {
      const newBrowseResults = { ...props.browseResults, total: undefined }
      expect(mapStateToProps({ ...props, browseResults: newBrowseResults }).total).toBe(0)
    })
  })

  describe('mapDispatchToProps', () => {
    describe('onFilterBoxChange', () => {
      it('should dispatch the `updateBrowseFilters` action with the correct params', () => {
        const change = { value: 123, type: 'test-type' }
        mapDispatchToProps(dispatch).onFilterBoxChange('category', change)
        expect(dispatch).toBeCalled()
        expect(updateBrowseFilters).toBeCalledWith('category', change.value, change.type)
      })
    })

    describe('onFilterTagClick', () => {
      it('should dispatch the `updateBrowseFilters` action with the correct params', () => {
        mapDispatchToProps(dispatch).onFilterTagClick('category', 'value')
        expect(dispatch).toBeCalled()
        expect(updateBrowseFilters).toBeCalledWith('category', 'value', 'remove')
      })
    })

    describe('onSortChange', () => {
      it('should dispatch the `updateBrowseParams` action with the correct params', () => {
        mapDispatchToProps(dispatch).onSortChange('value')
        expect(dispatch).toBeCalled()
        expect(updateBrowseParams).toBeCalledWith({ page: 1, sort: 'value' })
      })
    })

    describe('saveFilterData', () => {
      it('should dispatch the `saveFilterOptions` action', () => {
        mapDispatchToProps(dispatch).saveFilterData()
        expect(dispatch).toBeCalled()
        expect(saveFilterOptions).toBeCalled()
      })
    })

    describe('updatePage', () => {
      it('should dispatch the `updateBrowseParams` action', () => {
        mapDispatchToProps(dispatch).updatePage(123)
        expect(dispatch).toBeCalled()
        expect(updateBrowseParams).toBeCalledWith({ page: 123 })
      })
    })
  })
})
