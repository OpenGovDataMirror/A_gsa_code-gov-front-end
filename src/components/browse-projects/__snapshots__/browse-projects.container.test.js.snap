// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers - BrowseProjects mapStateToProps should return the correct properties 1`] = `
Object {
  "boxes": Object {
    "agencies": Array [
      Object {
        "checked": false,
        "name": "a1",
        "value": "v1",
      },
      Object {
        "checked": true,
        "name": "a2",
        "value": "v2",
      },
    ],
    "languages": Array [
      Object {
        "checked": true,
        "name": "l1",
        "value": "v3",
      },
    ],
    "licenses": Array [],
    "usageTypes": Array [],
  },
  "browseParams": Object {
    "filters": Array [
      Object {
        "category": "agencies",
        "modified": 1234567890123,
        "value": "v0",
      },
      Object {
        "category": "agencies",
        "modified": 1234567890123,
        "value": "v2",
      },
      Object {
        "category": "languages",
        "modified": 1234567890123,
        "value": "v3",
      },
    ],
    "page": "test-page",
    "size": "test-size",
    "sort": "test-sort",
  },
  "browseResults": Object {
    "repos": Array [
      Object {
        "repoId": "repo-1",
      },
      Object {
        "repoId": "repo-2",
      },
    ],
    "total": 2,
  },
  "filterTags": undefined,
  "repos": Array [
    Object {
      "repoId": "repo-1",
    },
    Object {
      "repoId": "repo-2",
    },
  ],
  "selectedPage": "test-page",
  "selectedPageSize": "test-size",
  "selectedSorting": "test-sort",
  "sortOptions": Array [
    Object {
      "label": "Data Quality",
      "selected": false,
      "value": "data_quality",
    },
    Object {
      "label": "A-Z",
      "selected": false,
      "value": "a-z",
    },
    Object {
      "label": "Last Updated",
      "selected": false,
      "value": "last_updated",
    },
  ],
  "total": 2,
}
`
